pkg/metrics/transport/transport.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transport
+
+import (
+	"context"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"k8s.io/klog/v2"
+)
+
+var (
+	/*
+		Proof that all labels are bound:
+		- method: known set of HTTP methods: GET, PUT, etc.
+		- service: restricted by extractService
+		- status_code: known set of HTTP status codes: 2xx, 3xx, etc.
+		- controller_name: known set of controller names, manually configured.
+	*/
+	APILabels = []string{"method", "service", "status_code", "controller_name"}
+
+	gcpRequestsTotal = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "configconnector",
+			Name:      "gcp_api_requests_total",
+			Help:      "Total number of GCP API requests",
+		},
+		APILabels,
+	)
+
+	gcpRequestDuration = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "configconnector",
+			Name:      "gcp_api_request_duration_seconds",
+			Help:      "Duration of GCP API requests in seconds",
+			Buckets:   prometheus.DefBuckets,
+		},
+		APILabels,
+	)
+
+	gcpErrorsTotal = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "configconnector",
+			Name:      "gcp_api_errors_total",
+			Help:      "Total number of GCP API errors",
+		},
+		APILabels,
+	)
+)
+
+type contextKey string
+
+const controllerNameContextKey contextKey = "controllerNameForMetrics"
+
+// WithControllerName returns a new context with the controller name set
+func WithControllerName(ctx context.Context, name string) context.Context {
+	return context.WithValue(ctx, controllerNameContextKey, name)
+}
+
+// wrap an http.RoundTripper to records metrics for all http requests
+type MetricsTransport struct {
+	inner http.RoundTripper
+}
+
+func NewMetricsTransport(inner http.RoundTripper) *MetricsTransport {
+	if inner == nil {
+		inner = http.DefaultTransport
+	}
+	return &MetricsTransport{inner: inner}
+}
+
+func (t *MetricsTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	service := extractService(req.URL.String())
+
+	start := time.Now()
+	resp, reqErr := t.inner.RoundTrip(req)
+	duration := time.Since(start).Seconds()
+
+	statusCode := "0" // good to have a default value here
+	if resp != nil {
+		statusCode = strconv.Itoa(resp.StatusCode)
```

Comment:
This is not bounded.

Reply:
I think we should be going through bounded types and then converting to string at the **WithLabelValues** call to ensure the labels are actually bounded.